src/v/storage/disk_log_impl.cc

```diff
@@ -557,43 +557,70 @@ model::term_id disk_log_impl::term() const {
     }
     return _segs.back()->offsets().term;
 }
+static offset_stats empty_offsets(model::offset start_offset) {
+    offset_stats ret;
+    ret.start_offset = start_offset;
+    if (ret.start_offset > model::offset(0)) {
+        ret.dirty_offset = ret.start_offset - model::offset(1);
+        ret.committed_offset = ret.dirty_offset;
+    }
+    return ret;
+}
 
 offset_stats disk_log_impl::offsets() const {
     if (_segs.empty()) {
-        offset_stats ret;
-        ret.start_offset = _start_offset;
-        if (ret.start_offset > model::offset(0)) {
-            ret.dirty_offset = ret.start_offset - model::offset(1);
-            ret.committed_offset = ret.dirty_offset;
-        }
-        return ret;
+        return empty_offsets(_start_offset);
     }
-    // NOTE: we have to do this because ss::circular_buffer<> does not provide
-    // with reverse iterators, so we manually find the iterator
+
     segment_set::type end;
     segment_set::type term_start;
+    segment_set::type last_committed;
+    /**
+     * We have to iterate backward over segments set as we have to find last
+     * non empty segment, last segment with initialized committed offset and
+     * first offset of current term
+     * NOTE: we have to do this because ss::circular_buffer<> does not
+     *       provide with reverse iterators, so we manually find the iterator
+     */
     for (int i = (int)_segs.size() - 1; i >= 0; --i) {
         auto& seg = _segs[i];
-        if (!seg->empty()) {
-            if (!end) {
-                end = seg;
-            }
-            // find term start offset
-            if (seg->offsets().term < end->offsets().term) {
+        // skip over empty segment
+        if (seg->empty()) {
+            continue;
+        }
+
+        // last not empty segment
+        if (!end) {
+            end = seg;
+        }
+
+        const auto& segment_offsets = seg->offsets();
+        // replace last_committed until we find segment with positive committed
+        // offset
```

Comment:
The loop also terminates when `term_start && segment_offsets.term < end->offsets().term`